src/client/sound_openal_internal.cpp

```diff
@@ -0,0 +1,1146 @@
+	s->cur_offset += copied_size;
+	return copied_size;
+}
+
+int OggVorbisBufferSource::seek_func(void *datasource, ogg_int64_t offset, int whence) noexcept
+{
+	OggVorbisBufferSource *s = (OggVorbisBufferSource *)datasource;
+	if (whence == SEEK_SET) {
+		if (offset < 0 || (size_t)offset > s->buf.size()) {
+			// offset out of bounds
+			return -1;
+		}
+		s->cur_offset = offset;
+		return 0;
+	} else if (whence == SEEK_CUR) {
+		if ((size_t)MYMIN(-offset, 0) > s->cur_offset
+				|| s->cur_offset + offset > s->buf.size()) {
+			// offset out of bounds
+			return -1;
+		}
+		s->cur_offset += offset;
+		return 0;
+	} else if (whence == SEEK_END) {
+		if (offset > 0 || (size_t)-offset > s->buf.size()) {
+			// offset out of bounds
+			return -1;
+		}
+		s->cur_offset = s->buf.size() - offset;
+		return 0;
+	}
+	return -1;
+}
+
+int OggVorbisBufferSource::close_func(void *datasource) noexcept
+{
+	std::unique_ptr<OggVorbisBufferSource> s((OggVorbisBufferSource *)datasource);
+	return 0;
+}
+
+long OggVorbisBufferSource::tell_func(void *datasource) noexcept
+{
+	OggVorbisBufferSource *s = (OggVorbisBufferSource *)datasource;
+	return s->cur_offset;
+}
+
+const ov_callbacks OggVorbisBufferSource::s_ov_callbacks = {
+	&OggVorbisBufferSource::read_func,
+	&OggVorbisBufferSource::seek_func,
+	&OggVorbisBufferSource::close_func,
+	&OggVorbisBufferSource::tell_func
+};
+
+/*
+ * RAIIOggFile struct
+ */
+
+Optional<OggFileDecodeInfo> RAIIOggFile::getDecodeInfo(const std::string &filename_for_logging)
+{
+	OggFileDecodeInfo ret;
+
+	vorbis_info *pInfo = ov_info(&m_file, -1);
+	if (!pInfo)
+		return nullopt;
+
+	ret.name_for_logging = filename_for_logging;
+
+	if (pInfo->channels == 1) {
+		ret.is_stereo = false;
+		ret.format = AL_FORMAT_MONO16;
+		ret.bytes_per_sample = 2;
+	} else if (pInfo->channels == 2) {
+		ret.is_stereo = true;
+		ret.format = AL_FORMAT_STEREO16;
+		ret.bytes_per_sample = 4;
+	} else {
+		warningstream << "Audio: Can't decode. Sound is neither mono nor stereo: "
+				<< ret.name_for_logging << std::endl;
+		return nullopt;
+	}
+
+	ret.freq = pInfo->rate;
+
+	ret.length_samples = static_cast<ALuint>(ov_pcm_total(&m_file, -1));
+	ret.length_seconds = static_cast<f32>(ov_time_total(&m_file, -1));
+
+	return ret;
+}
+
+RAIIALSoundBuffer RAIIOggFile::loadBuffer(const OggFileDecodeInfo &decode_info,
+		ALuint pcm_start, ALuint pcm_end)
+{
+	constexpr int endian = 0; // 0 for Little-Endian, 1 for Big-Endian
+	constexpr int word_size = 2; // we use s16 samples
+	constexpr int word_signed = 1; // ^
+
+	// seek
+	if (ov_pcm_tell(&m_file) != pcm_start) {
+		if (ov_pcm_seek(&m_file, pcm_start) != 0) {
+			warningstream << "Audio: Error decoding (could not seek) "
+					<< decode_info.name_for_logging << std::endl;
+			return RAIIALSoundBuffer();
+		}
+	}
+
+	const size_t size = static_cast<size_t>(pcm_end - pcm_start)
+			* decode_info.bytes_per_sample;
+
+	std::unique_ptr<char[]> snd_buffer(new char[size]);
```

Comment:
I suggest just using a `std::vector`

Reply:
Afaik, it is not easily possible to make a std::vector with uninitialized elements. So that would be slightly slower (and hence less elegant).

This here on the other hand does the same as C++17 std::make_unique_for override.

Reply:
Well whatever.